plugin-server/src/worker/vm/upgrades/historical-export/export-historical-events-v2.ts

```diff
@@ -0,0 +1,487 @@
+    // Only set starting export from UI
+    parallelism?: number
+    dateFrom?: string
+    dateTo?: string
+}
+
+export interface ExportParams {
+    id: number
+    parallelism: number
+    dateFrom: ISOTimestamp
+    dateTo: ISOTimestamp
+}
+
+interface CoordinationPayload {
+    running?: Array<ISOTimestamp>
+    done?: Array<ISOTimestamp>
+    progress?: number
+}
+
+interface CoordinationUpdate {
+    hasChanges: boolean
+    done: Array<ISOTimestamp>
+    running: Array<ISOTimestamp>
+    toStartRunning: Array<[ISOTimestamp, ISOTimestamp]>
+    progress: number
+    exportIsDone: boolean
+}
+
+interface ExportDateStatus {
+    done: boolean
+    progress: number
+}
+
+export function addHistoricalEventsExportCapabilityV2(
+    hub: Hub,
+    pluginConfig: PluginConfig,
+    response: PluginConfigVMInternalResponse<PluginMeta<ExportHistoricalEventsUpgradeV2>>
+): void {
+    const { methods, tasks, meta } = response
+
+    const currentPublicJobs = pluginConfig.plugin?.public_jobs || {}
+
+    if (!(INTERFACE_JOB_NAME in currentPublicJobs)) {
+        hub.promiseManager.trackPromise(hub.db.addOrUpdatePublicJob(pluginConfig.plugin_id, INTERFACE_JOB_NAME, {}))
+    }
+
+    const oldRunEveryMinute = tasks.schedule.runEveryMinute?.exec
+
+    tasks.job[INTERFACE_JOB_NAME] = {
+        name: INTERFACE_JOB_NAME,
+        type: PluginTaskType.Job,
+        // TODO: Accept timestamp as payload
+        exec: async (payload: ExportHistoricalEventsUIPayload) => {
+            // only let one export run at a time
+            const alreadyRunningExport = await getExportParameters()
+            if (!!alreadyRunningExport) {
+                return
+            }
+
+            // :TODO: Clear/invalidate old export storage somehow
+
+            const id = Math.floor(Math.random() * 10000 + 1)
+            const parallelism = Number(payload.parallelism ?? 1)
+            const boundaries = await getTimestampBoundaries(payload)
+
+            await meta.storage.set(EXPORT_RUNNING_KEY, {
+                id,
+                parallelism,
+                dateFrom: boundaries.min.toISOString(),
+                dateTo: boundaries.max.toISOString(),
+            } as ExportParams)
+
+            await coordinateHistoricExport()
+        },
+    } as unknown as PluginTask // :KLUDGE: Work around typing limitations
+
+    tasks.job['exportHistoricalEvents'] = {
+        name: 'exportHistoricalEvents',
+        type: PluginTaskType.Job,
+        exec: (payload) => exportHistoricalEvents(payload as ExportHistoricalEventsJobPayload),
+    }
+
+    tasks.schedule.runEveryMinute = {
+        name: 'runEveryMinute',
+        type: PluginTaskType.Schedule,
+        exec: async () => {
+            await oldRunEveryMinute?.()
+            await coordinateHistoricExport()
+        },
+    }
+
+    async function exportHistoricalEvents(payload: ExportHistoricalEventsJobPayload): Promise<void> {
+        if (payload.retriesPerformedSoFar >= 15) {
```

Comment:
Let's make this number configurable

Reply:
Configurable by whom?

Reply:
Make this an env var

Reply:
Will do in a separate PR as a follow-up.